timing/python/framework/benchmark.py

```diff
@@ -0,0 +1,290 @@
+
+    def __init__(self, time, cpu, num_iter):
+        self.time = float(time)
+        self.cpu = float(cpu)
+        self.num_iter = num_iter
+
+    def get_time(self, unit="us"):
+        return self.time * 1e6
+
+    def get_cpu(self, unit="us"):
+        return self.cpu * 1e6
+
+    def get_num_iter(self):
+        """
+        Number of iterations run to get timings
+        """
+        return self.num_iter
+
+    def __repr__(self):
+        return ", ".join(k + "=" + str(v) for k, v in self.items())
+
+
+class BenchmarkResults(object):
+    def __init__(self):
+        self.results = []
+
+    def time_mean(self, unit="us"):
+        return int(
+            np.mean(list(map(lambda x: x.get_time(unit), self.results)))
+        )
+
+    def time_std(self, unit="us"):
+        return int(np.std(list(map(lambda x: x.get_time(unit), self.results))))
+
+    def cpu_mean(self, unit="us"):
+        return int(np.mean(list(map(lambda x: x.get_cpu(unit), self.results))))
+
+    def cpu_std(self, unit="us"):
+        return int(np.std(list(map(lambda x: x.get_cpu(unit), self.results))))
+
+    def iter_mean(self):
+        return int(
+            np.mean(list(map(lambda x: x.get_num_iter(), self.results)))
+        )
+
+    def append(self, result):
+        assert isinstance(result, BenchmarkResult)
+        self.results.append(result)
+
+
+class Benchmark(object):
+    def __init__(self):
+        if "args" not in dir(self):
+            logging.warn(
+                "Benchmark " + str(self.__class__) + " has no args set"
+            )
+            return
+        cpus = bench_utils.get_cpu_list()
+        for cpu in cpus:
+            bench_utils.check_cpu_governor(cpu)
+        self.state = {}
+
+
+class ListBenchmark(Benchmark):
+    """
+    Basic benchmark class. Expects list of arguments and processes one at a
+    time.
+    """
+
+    def __init__(self):
+        super(ListBenchmark, self).__init__()
+        if not isinstance(self.args, list):
+            raise TypeError("args needs to be a list of arguments")
+
+
+class GridBenchmark(Benchmark):
+    """
+    Creates a grid of arguments and calls benchmark with each.
+    The arguments must be primited types such as strings or
+    numbers and will be shallow(!) copied as part of the setup
+    """
+
+    def __init__(self):
+        super(GridBenchmark, self).__init__()
+        if not isinstance(self.args, dict):
+            raise TypeError("args needs to be a dict of arguments")
+        self.args = bench_utils.grid(self.args)
+
+
+# TODO
+# benchmark [--benchmark_list_tests={true|false}]
+#           [--benchmark_filter=<regex>]
+#           [--benchmark_format=<console|json|csv>]
+#           [--benchmark_out=<filename>]
+#           [--benchmark_out_format=<json|console|csv>]
+#           [--benchmark_color={auto|true|false}]
+#           [--benchmark_counters_tabular={true|false}]
+#           [--v=<verbosity>]
+# DONE
+#           [--benchmark_min_time=<min_time>]
```

Comment:
Not sure where to leave this comment, but what units are benchmark_min_time and benchmark_max_time in?

Reply:
seconds. we still need user counters etc. I'm only leaving it for myself. I guess I can remove it. I expect another PR tonight that does 50% of this.